### PR TITLE
Feature/blast mine plugin

### PR DESCRIPTION
**Pull request by PandahRS**
Follow up on the leg end Unmoon's PR #1079.
Code has been based on the code review and an experience counter for ores within the ore sack.

[Example here](https://streamable.com/t2yj7)

### CODE REVIEW

**Comment by Jalopyy** [comment]
If I can suggest one more thing, how about a line on the window that has gold value on collect? IE the sack is worth X gp.

**Comment by PandahRS** [comment]
@Jalopyy I'm not sure if I can add this in the same window as well. It'd have to be seperated with the collected xp. I'll look into it though.

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineOreDetailsOverlay.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	{
+		// convert to unnoted id
+		ItemComposition composition = this.itemManager.getItemComposition(itemId);
+		final boolean note = composition.getNote() != -1;
+		final int id = note ? composition.getLinkedNoteId() : composition.getId();
+
+		ItemPrice itemPrice;
+		try
+		{
+			itemPrice = this.itemManager.getItemPrice(id);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineOreDetailsOverlay.java]: Don't do blocking item price lookups on the game thread.

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	{
+		return Arrays.asList(blastMineRockOverlay, blastMineOreCountOverlay, blastMineOreDetailsOverlay);
+	}
+
+	@Subscribe
+	public void onGameObjectSpawned(GameObjectSpawned event)
+	{
+		GameObject gameObject = event.getGameObject();
+
+		BlastMineRock newRock = new BlastMineRock(gameObject);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java]: This should check the object id before allocating a blastminerock around it

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+			rocks.add(newRock);
+		}
+	}
+
+	@Subscribe
+	public void onGameStateChanged(GameStateChanged event)
+	{
+		if (event.getGameState() == GameState.LOADING)
+		{
+			rocks.clear();
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java]: Also game object despair event would probably be useful

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+import java.util.Collections;
+import java.util.Set;
+
+public class BlastMineRockOverlay extends Overlay
+{
+	private static final int TIMER_SIZE = 25;
+	private static final int TIMER_BORDER_WIDTH = 1;
+	private static final int MAX_DISTANCE = 16;
+	private static final int WARNING_DISTANCE = 2;
+	private static final Set<Integer> WALL_OBJECTS = Collections.unmodifiableSet(Sets.newHashSet(ObjectID.NULL_28570, ObjectID.NULL_28571, ObjectID.NULL_28572, ObjectID.NULL_28573, ObjectID.NULL_28574,
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java]: `ImmutableSet`

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	private final BlastMinePlugin plugin;
+	private final BlastMinePluginConfig config;
+	private final ItemManager itemManager;
+
+	private Color timerColor;
+	private Color timerBorderColor;
+	private Color warningColor;
+
+	@Inject
+	BlastMineRockOverlay(@Nullable Client client, BlastMinePlugin plugin, BlastMinePluginConfig config, ItemManager itemManager)
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java]: Client is not nullable unless the plugin is configured to load when outdated, which it should not be.

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+		this.plugin = plugin;
+		this.config = config;
+		this.client = client;
+		this.itemManager = itemManager;
+	}
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		plugin.getRocks().removeIf(rock -> (rock.getRemainingTimeRelative() == 1 && rock.getType() != BlastMineRockType.NORMAL) || (rock.getRemainingFuseTimeRelative() == 1 && rock.getType() == BlastMineRockType.LIT));
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java]: Removal of rocks should be on game tick probably

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	{
+		timerColor = config.getTimerFillColor();
+		timerBorderColor = config.getTimerBorderColor();
+		warningColor = config.getWarningColor();
+	}
+
+	private void drawRocks(Graphics2D graphics)
+	{
+		Tile[][][] tiles = client.getRegion().getTiles();
+		BufferedImage chiselIcon = itemManager.getImage(ItemID.CHISEL);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java]: Avoid extra work by checking if there are rocks nearby first ..

Comment (Unmoon) [reply runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMineRockOverlay.java]: From previous review, icons can be fetched once and re-used each time.

File: runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java (hunk 3 above)
Comment (PandahRS) [reply runelite-client/src/main/java/net/runelite/client/plugins/blastmine/BlastMinePlugin.java]: Would you want to check anything within the `onGameObjectDespawned` or event or should it function as a simple clear(). as above?

**Comment by PandahRS** [comment]
Adjustments are made in the last commit based on the code review.

**Pull request closed without merging**